File: leetcode/subsets-ii.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums)
    {
        set<vector<int>> res;
        vector<int> cur;

        auto length = nums.size();

        sort(nums.begin(), nums.end());

        function<void(int, vector<int>&)> backtrack = [&](int l,
                                                          vector<int> cur) {
            res.insert(cur);

            if (l >= length)
                return;

            for (auto i = l; i < length; ++i) {
                // take i
                cur.push_back(nums[i]);

                backtrack(i + 1, cur);

                cur.pop_back();
            }
        };

        backtrack(0, cur);

        return vector<vector<int>>(res.begin(), res.end());
    }
};

File: leetcode/subsets-ii.py
from typing import List


class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        length = len(nums)

        res = []

        nums.sort()

        def backtrack(l: int, cur: List[int]):
            res.append(cur.copy())
            if l >= length:
                return

            for i in range(l, length):
                if i < l + 1 or nums[i] != nums[i - 1]:
                    cur.append(nums[i])
                    backtrack(i + 1, cur)
                    cur.pop()

        backtrack(0, [])

        return res
